### PR TITLE
Resurrect type and shape

### PR DESCRIPTION
**Pull request by yuanbyu**
This introduced the type and shape annotation into ONNX. For the inputs and outputs of the graph, we would like to make it required to specify both type and shape. This allows us to state the contract of the model more precisely.

The ShapeProto introduces symbolic dimension. This is useful to state the shape contract properly. For example, if a model has two matrix inputs X and Y and requires that the dim 2 of X must be the same as the dim 1 of Y.  We can specify X.shape = [m, n] and Y.shape = [n, k].

### CODE REVIEW

**Comment by yuanbyu** [comment]
a) In GraphProto, ArgInfoProto is repeated and we don't require its presence, so it is optional. 
b) We support protobuf v2.6.1 and up. So one-of is allowed but map is not allowed.

**Comment by donbox** [comment]
Yuan, your first comment is at odds with your last comment.

I believe that declaration of input and output types is required on graph (required for type inference) and but we CAN live with it optional on node assuming the implementation has knowledge of the declared signature for the referenced operator.

**Comment by yuanbyu** [comment]
Oops it was a typo. It should have said "In **NodeProto**, ArgInfoProto is repeated ...". Thanks Don for pointing it out.

File: onnx/onnx.proto [hunk 1]
```diff
@@ -212,6 +237,44 @@ message TensorProto {
...
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  message SparseTensorTypeProto {
+  // This field MUST NOT have the value of UNDEFINED
+  // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  oneof value {
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: Why do we need a separation (especially of "oneof" kind) for sparse vs dense tensors? Element type is present in either Tensor or SparseTensor.values. Maybe put TensorProto.DataType as member of ArgInfoProto directly?

Or do you imagine some other fields appearing in TypeProto struct?

File: onnx/onnx.proto [hunk 2]
```diff
@@ -212,6 +237,44 @@ message TensorProto {
...
+
+    // The type of a sparse tensor.
+    SparseTensorTypeProto sparse_tensor_type = 2;
+  }
+}
+
+// Defines a tensor shape. A dimension can be either an integer value
+// or a symbolic variable. A symbolic variable represents an unknown
+// dimension.
+message TensorShapeProto {
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: another (less descriptive) option would be to use negative values in "shape" as symbolic references, e.g. "-1", "-2", etc. It'd be less descriptive but would simplify the format and use the same `repeated int64 dims` for both fully-realized shape and partially specified shape.

File: onnx/onnx.proto [hunk 3]
```diff
@@ -212,6 +237,44 @@ message TensorProto {
...
+  message Dimension {
+    oneof value {
+      int64 dim_value = 1;
+      string dim_param = 2;   // namespace Shape
+    };
+  };
+  repeated Dimension dim = 1;
+}
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: Should we also support arbitrary list of dimensions? E.g. we can have "named range of dimensions". One option to specify it would be to put `string dim_list` above that can cover 0 or more arbitrary dimensions preserved "as is".

For example `Gather` operator would be taking shapes `[N, ...dims...]` and `[K]` and returning `[K, ...dims...]`.

File: onnx/onnx.proto [hunk 4]
```diff
@@ -64,21 +65,34 @@ message AttributeProto {
...
+  optional string name = 3;     // namespace Node
+  optional string op_type = 4;  // namespace Operator
 
   // Additional named attributes.
   repeated AttributeProto attribute = 5;
 
   // A human-readable documentation for this graph. Markdown is allowed.
   optional string doc_string = 6;
+
+  // Information including type and shape for each input and output of
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: I'm a little confused what is the role of specification at the individual Node level. In particular all inputs of the node are some other Node's outputs, so the type/shape information is redundant. If we wanted to preserve types and shapes of intermediate value it feel more natural to just record them in "value_name -> ArgInfoProto" directly in the graph.

Second, how does it connect with the type specifications and inferences #55 ? If we're moving towards full automatic shape inference there is little need to record the shapes of the intermediate values in the model proto.

Comment (donbox) [reply onnx/onnx.proto]: On your first point, you are correct that the type declarations on Node require redundant specification on both the input and output side. If we keep this rep, we need to define the rules for what happens when they TypeProto is not identical.   I'll defer to Yuan as to whether encoding a single table that declares the type of each edge/named-value is better or worse.

On your second point, one could take a graph proto and the library protos for all referenced operators and either (a) fill in absent Node type decls or (b) check the ones that are there.  Note that in Yuan's PR, he's only requiring type decls on Graph input/output, not Node (which as you state can be inferred).

File: onnx/onnx.proto (hunk 1 above)
Comment (donbox) [reply onnx/onnx.proto]: See #55 for other TypeProto variants, which are needed to properly schematize attribute values (which can be scalars or sequences, not just tensors)

.

File: onnx/onnx.proto (hunk 2 above)
Comment (donbox) [reply onnx/onnx.proto]: That would work, however using symbolic names has two advantages:

1. It coins a name one can refer to in doc_string.

2. If the same name appears two or more times on a given graph or node, it's value must be the same, which addresses correlating dimensions across input and output parameters..

File: onnx/onnx.proto (hunk 3 above)
Comment (donbox) [reply onnx/onnx.proto]: I've hit the need for something like this as well. If I understand correctly, you are saying we'd effectively add a rest arg/variatic arg and allow it to be named, again, to support correlation?

File: onnx/onnx.proto (hunk 4 above)
Comment (gramalingam) [reply onnx/onnx.proto]: Agree with Don. On the second point, this allows the possibility of a type-inference being a stand-alone tool, that takes a graph-proto as input and writes a (fully type-annotated) graph-proto as output.

Comment (dzhulgakov) [reply onnx/onnx.proto]: On the second point - what is the use case of recording intermediate tensor shapes in the serialized model? I'm a bit biased towards keeping the minimum necessary functionality for model interchange in ONNX whereas inference of shapes can be handled by corresponding frameworks. As you said, we can always infer shapes based on Node definitions, operator specs and graph inputs.

File: onnx/onnx.proto (hunk 1 above)
Comment (yuanbyu) [reply onnx/onnx.proto]: Yes, TypeProto is an extension point, and we would like to have the flexibility to extend it t introduce more data types in the future.

File: onnx/onnx.proto (hunk 2 above)
Comment (dzhulgakov) [reply onnx/onnx.proto]: Agree, symbolic names are indeed more readable. Probably it's fine at expense of protobuf brevity.

File: onnx/onnx.proto (hunk 3 above)
Comment (dzhulgakov) [reply onnx/onnx.proto]: Yes, adding a "variadic placeholder" that can capture zero or more dimensions

File: onnx/onnx.proto (hunk 2 above)
Comment (yuanbyu) [reply onnx/onnx.proto]: It could also be used for more efficient memory reuse, when we for example statically plan for the memory allocation/reuse.

Comment (yuanbyu) [reply onnx/onnx.proto]: Yes, using negative numbers would work, but somewhat less elegant in my opinion.

File: onnx/onnx.proto (hunk 4 above)
Comment (yuanbyu) [reply onnx/onnx.proto]: On the first point, I am fine with encoding them as "value_name -> ArgInfoProto" in graph. 

On the second point, I don't think that this PR and Don's PR commit to fully automatic shape inference. We can discuss if that is the direction we should go and the issues we might run into.
For example, do we really want to require all participating frameworks to perform fully automatic shape inference? Even for type inference, while we will definitely d it, it is uncertain to me if we want to require it for all frameworks.

File: onnx/onnx.proto (hunk 3 above)
Comment (yuanbyu) [reply onnx/onnx.proto]: TF has a more general `Gather` where the axis to gather is an input. Do we want to handle it?

**Comment by ezyang** [comment]
I apologize if this has been written elsewhere, but it would be really helpful to know what use-cases we expect backends to take advantage of this type information for. For example, Caffe2 cannot make use of this type information at all, because it always waits until it knows what the concrete inputs are, at which point it can recompute all of the information. Do we have any short term use cases where this truly is untenable?

File: onnx/onnx.proto (hunk 4 above)
Comment (jerryzh168) [reply onnx/onnx.proto]: For type/shape inference, is it possible to do it in the onnx level so that participating frameworks don't need to implement it again?

Comment (jerryzh168) [reply onnx/onnx.proto]: We can do something like:
1. provide a set of default type/shape inference functions
2. offer a way to override the default inference rules.

File: onnx/onnx.proto [hunk 5]
```diff
@@ -124,7 +139,17 @@ message GraphProto {
 
   // A human-readable documentation for this graph. Markdown is allowed.
   optional string doc_string = 10;
-};
+
+  // Argument information for input/output of this graph.
+  // Each ArgInfoProto corresponds one-to-one by position with the
```
Comment (gramalingam) [inline onnx/onnx.proto]: If we are willing to do a BC breaking change, it would be better to bundle information about an input (or output) together: name, type, shape, and even default-values (which are currently in the initializer list), instead of correlating across lists by position or name etc. (with the possibility of missing or mismatching entries).

File: onnx/onnx.proto (hunk 3 above)
Comment (dzhulgakov) [reply onnx/onnx.proto]: It's actually a perfect example why static shape inference might be limiting and providing C++ inference function is beneficial. E.g. that's exactly what TF does in this case: https://github.com/tensorflow/tensorflow/blob/3c0bcf8442a107b635cca64440d05c390a35d177/tensorflow/core/ops/array_ops.cc#L1396

I think for now we can avoid "vararg" expressions. For inputs/outputs specification even simple symbolic names like in this PR should be sufficient.

File: onnx/onnx.proto (hunk 4 above)
Comment (donbox) [reply onnx/onnx.proto]: Dmytro, we should definitely collapse the GraphProto.input/GraphProto.input_arg_info fields for both compactness and to make it less likely that folks don't screw it up..

I suspect once we write down the type inference rules, someone will contribute a reference implementation. I agree that would be pure goodness and will help us on interop.

File: onnx/onnx.proto (hunk 5 above)
Comment (donbox) [reply onnx/onnx.proto]: I do have a strong preference to at least put the arginfo and name stuff together.

I'm less convicted on putting initializer in there given the other threads going on. 

Ed?

**Comment by donbox** [comment]
We have scenarios where we will import a model and deploy it with an app, and that the type/shape of inputs is statically known.

The ability to do a validity check at model import against the known signature of operators helps us catch bad models earlier in the developer work flow (hence the desire for the operator library stuff).

The ability to cache the results of type/shape inference into the rep directly (hence the desire for putting type/shape annotation on Node) reduces load time as we know we’ve already checked the model.  On client configurations, load time matters. 

DB



Sent from Mail for Windows 10
[16299 - 16.0.8600.4040]

**Comment by yuanbyu** [comment]
@dzhulgakov Based on your suggestion, I moved the type and shape out of NodeProto and into GraphProto. Please take another look?

File: onnx/onnx.proto [hunk 6]
```diff
@@ -64,15 +65,25 @@ message AttributeProto {
   repeated GraphProto graphs = 11;    // list of graph
 }
 
+// Defines information on value, including the name, the type, and
+// the shape of the value.
+message ValueInfoProto {
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: this would be a breaking change for PyTorch which spits out the proto in C++: https://github.com/pytorch/pytorch/blob/5deacb5bce8bfb41662b57a61ea635c44d2a347f/torch/csrc/onnx/onnx.h#L249

Just FYI for @ezyang

File: onnx/onnx.proto [hunk 7]
```diff
@@ -64,15 +65,25 @@ message AttributeProto {
...
+  // This field MUST be present in this version of the IR.  
+  optional string name = 1;     // namespace Value
+  // This field MUST be present in this version of the IR.
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: I think it MUST be present only if it's graph inputs/outputs. Let's clarify that if it's used as a part of internal values definitions (`value_info` below) then it's fine to have type optional

File: onnx/onnx.proto [hunk 8]
```diff
@@ -124,7 +136,11 @@ message GraphProto {
 
   // A human-readable documentation for this graph. Markdown is allowed.
   optional string doc_string = 10;
-};
+
+  // Information for the values in the graph. The ValueInfoProto.name's
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: describe that it's optional?

File: onnx/onnx.proto [hunk 9]
```diff
@@ -103,6 +114,7 @@ message GraphProto {
   repeated TensorProto initializer = 5;
 
   // The version of the IR this graph targets. See Version enum below.
+  // This field MUST be present this version of the IR.
   optional int64 ir_version = 6;
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: we should probably bump IR version because it's indeed a breaking change (just following the protocol)

File: onnx/onnx.proto [hunk 10]
```diff
@@ -92,8 +103,8 @@ message GraphProto {
   optional string name = 2;   // namespace Graph
 
   // The inputs and outputs of the graph.
-  repeated string input = 3;  // namespace Tensor
-  repeated string output = 4; // namespace Tensor
+  repeated ValueInfoProto input = 3;  // namespace Value
```
Comment (dzhulgakov) [inline onnx/onnx.proto]: as we discussed in a chat - this is not safe. Maybe add fields with another id?

Comment (donbox) [reply onnx/onnx.proto]: I think we should just bump the IR version and take this change along with #53 as a single BC.

If we get this to master this week, I think we are fine.

File: onnx/onnx.proto (hunk 9 above)
Comment (donbox) [reply onnx/onnx.proto]: Yup. See my last comment.

File: onnx/onnx.proto (hunk 7 above)
Comment (yuanbyu) [reply onnx/onnx.proto]: It is optional for a value to appear in `value_info`. So when a value appears in `value_info`, the PR currently does require type.  Do you think we should relax it further in this case?

File: onnx/onnx.proto (hunk 9 above)
Comment (yuanbyu) [reply onnx/onnx.proto]: Will do.

File: onnx/onnx.proto (hunk 8 above)
Comment (yuanbyu) [reply onnx/onnx.proto]: Done.

File: onnx/onnx.proto (hunk 10 above)
Comment (yuanbyu) [reply onnx/onnx.proto]: It certainly makes things less clean. I will assign new ids so we can look at it again.

**Comment by ezyang** [comment]
@fmassa has been working on a TensorFlow importer for ONNX, and the TensorFlow frontend *requires* shape and type information, so we'll need this for that.

**Comment by fmassa** [comment]
I just found out that TF requires `type` information, but the `shape` is optional.

File: onnx/onnx.proto [hunk 11]
```diff
@@ -75,15 +76,25 @@ message AttributeProto {
...
 }
 
+// Defines information on value, including the name, the type, and
+// the shape of the value.
+message ValueInfoProto {
+  // This field MUST be present in this version of the IR.  
+  optional string name = 1;     // namespace Value
+  // This field MUST be present in this version of the IR.  
+  optional TypeProto type = 2;
+  optional TensorShapeProto shape = 3;
```
Comment (ezyang) [inline onnx/onnx.proto]: I have been thinking about this some more, and I don't think shape should be besides the TypeProto. The shape specification can depend on whether or not you have a dense or sparse tensor; for example, PyTorch sparse tensors have both sparse dimensions and dense dimensions, and that should be required in the shape.

File: onnx/onnx.proto [hunk 12]
```diff
@@ -37,11 +37,12 @@ package onnx;
 // - Attribute: These names identify attribute names for extra attributes that
 //   are passed to operators.
 // - Operator: These names identify particular operators.
-// - Tensor: These names identify intermediate tensor values flowing through
+// - Value: These names identify intermediate values (typically tensors) flowing through
```
Comment (ezyang) [inline onnx/onnx.proto]: Whether or not we actually rename Tensor to Value is, I think, blocked on figuring out what happens in https://github.com/onnx/onnx/pull/67

File: onnx/helper.py [hunk 13]
```diff
@@ -31,8 +31,10 @@ def make_graph(nodes, name, inputs, outputs, initializer=[]):
     graph = GraphProto()
     graph.node.extend(nodes)
     graph.name = name
-    graph.input.extend(inputs)
-    graph.output.extend(outputs)
+    for x in inputs:
```
Comment (bddppq) [inline onnx/helper.py]: Since we have decided to make `type` and `shape` as required in the graph input and output, the argument `inputs` and `outputs` in this `make_graph` function should not be a list of string anymore, otherwise, this `make_graph` will actually produce an invalid graph.

Comment (yuanbyu) [reply onnx/helper.py]: Good point. Any proposal how we should change it?

Comment (bddppq) [reply onnx/helper.py]: inputs and outputs should be a list of corresponding proto structs (like initializer), so the original code is fine (you just need to revert the changes in this file).

Comment (yuanbyu) [reply onnx/helper.py]: OK. I will add a MakeType() method to make it easier to construct.

Comment (yuanbyu) [reply onnx/helper.py]: @bddppq Thanks for fixing my bug. I was going to look at after dinner, but pleasantly found it fixed.
